### PR TITLE
Support yaml streams when decoding

### PR DESCRIPTION
**Pull request by chancez**
A port of my PR at the upstream repo to support yaml streams with the unmarshal methods: https://github.com/ghodss/yaml/pull/47

### CODE REVIEW

**Comment by dims** [comment]
@chancez do you mind throwing this change at the vendored copy of kubernetes/kubernetes and see what breaks if any? (just a WIP with a patched yaml.go in main kubernetes repo)

**Comment by chancez** [comment]
@dims as a PR? sure

**Comment by dims** [comment]
@chancez yes! so we can see if the CI catches any issues

**Comment by chancez** [comment]
https://github.com/kubernetes/kubernetes/pull/73227 is up

**Comment by chancez** [comment]
Got some failures in kube/kube...looking into it now.

**Comment by chancez** [comment]
Ok, I found a way to reproduce one error (the panic) from kube/kube and it's related to unmarshalling a yaml document with the content `null`. Currently adding a test for it and debugging. It's reflection related.

**Comment by dims** [comment]
very cool! thanks @chancez

**Comment by chancez** [comment]
Okay, I fixed the panic I saw and added some tests for it. I'm looking into the other failures in kube/kube. I believe it's likely related to kube/kube having it's own logic/wrappers to handle streaming, but that's just a guess.

**Comment by chancez** [comment]
Ah, how interesting, at least one kube/kube failure is that my code produces an `[]` when given `""` as the input. Before my change, it outputs `null`. The json unmarshaller behavior seems similar so that seems correct.

**Comment by chancez** [comment]
Okay, I believe I fixed the kube/kube bazel test failures. I just pushed up with the updates I made in to this PR and it's re-running tests.

**Comment by chancez** [comment]
Kube/kube is now only failing on godep and kops.

**Comment by dims** [comment]
@chancez  looks like you worked through some thorny stuff which would have bitten us. nice! thanks.

**Comment by dims** [comment]
@deads2k @sttts @lavalamp one of you wanna take a look? thanks!

File: yaml.go [hunk 1]
```diff
@@ -96,6 +96,62 @@ func JSONToYAML(j []byte) ([]byte, error) {
...
+// internalYamlUnmarshalStrict works like yaml.UnmarshalStrict but supports streams.
+func internalYamlUnmarshalStrict(b []byte, out interface{}) error {
+	return internalYamlUnmarshalWrapper(true, b, out)
+}
+
+// internalYamlUnmarshalWrapper is a wrapper around yaml.Decoder that allows
+// for decoding yaml streams.
+func internalYamlUnmarshalWrapper(strict bool, b []byte, out interface{}) error {
+	if len(bytes.TrimSpace(b)) == 0 {
+		out = nil
```
Comment (chancez) [inline yaml.go]: I realized this has no effect and I still need to use reflect here and down on 144 too. Sending a new commit for that.

**Comment by chancez** [comment]
Squashed and made a minor fix to correctly set `out` using reflection when the input value is empty/null. I'm looking into benchmarking now. Do we want these to be just regular go test benchmarks? Trying to figure out what a good dataset would even be.

File: yaml.go [hunk 2]
```diff
@@ -96,6 +96,61 @@ func JSONToYAML(j []byte) ([]byte, error) {
...
+		err := decoder.Decode(&tmp)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+		outs = append(outs, tmp)
+	}
+
+	if len(outs) == 0 {
```
Comment (neolit123) [inline yaml.go]: this branch is not needed as a no-op.

File: yaml.go [hunk 3]
```diff
@@ -96,6 +96,61 @@ func JSONToYAML(j []byte) ([]byte, error) {
...
+	} else if len(outs) == 1 {
+		// if there is only one value, we don't want to put it into a slice
+		item := outs[0]
+		// the we got null as the input
+		if item == nil {
+			outV.Set(reflect.Zero(outV.Type()))
+			return nil
+		}
+		outV.Set(reflect.ValueOf(item))
+	} else {
```
Comment (neolit123) [inline yaml.go]: handle `>0` here?

Comment (neolit123) [reply yaml.go]: adding this here `if len(outs) > 1 {` and removing the `if len(outs) == 0 {` branch.

**Comment by chancez** [comment]
I'm realizing that in the null/empty input cases I'm not overwriting `out` to be it's zero value/nil, and that seemed incorrect to me, but it seems similar to how the JSON unmarshaller behaves: https://play.golang.org/p/a5s0Z1md9L7

The only one im unsure of is what should be doing when:
- `out` is a non-zero value, aka: it could be a map with items in it.
- the input is an empty yaml stream: `---\n`

I've added a test for the other cases I tested in the playground link and I'm working on making the behavior pass the tests.

File: yaml.go (hunk 2 above)
Comment (neolit123) [reply yaml.go]: i will leave this question to the api-machinery folks, but this makes sense to me:
> I think you would expect null, so I think I need to set out to zero in this branch.

File: yaml.go (hunk 3 above)
Comment (chancez) [reply yaml.go]: The `else` already is covering the ` > 1` in either case.

Comment (neolit123) [reply yaml.go]: if `if len(outs) == 0` is removed and if it remains a no-op, this branch has to handle `> 1`.

**Comment by chancez** [comment]
I'm stuggling to figure out how to make https://github.com/kubernetes-sigs/yaml/pull/12/commits/b3c570ac403e5c01b56ab20c2ba655ac3ebd2ad5#diff-b9712b0e0447a2be5d319fb286dee2afR152 pass. Any tips would be good.

**Comment by chancez** [comment]
Well, I've actually hit an even bigger issue in the reflection code that makes me realize the reflection here isn't correctly handling the `out` parameter's underlying type.  While working on benchmarking, I found this error:

```
panic: reflect.Set: value of type []interface {} is not assignable to type []*yaml.benchmarkTestItem
```

I understand what I need to do, just noting it here.

I'm also updating the benchmark to test `yaml.Unmarshal` and `internalYamlUnmarshal` directly so we can have both run as part of our tests to get a baseline.

**Comment by neolit123** [comment]
> panic: reflect.Set

i was going to ask if there is potential for panics in `Set`... :)

**Comment by chancez** [comment]
I'm going to close this for now until I come up with a better solution/idea.

**Pull request closed without merging**